### ISSUE TITLE
[tf2.0.0] tf.keras.layers.GRU incorrect output of model.fit_generator trying to run Francois Chollet's notebook

### ISSUE THREAD

**Issue by dbonner** [tensorflow/tensorflow#32987]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): I have slightly modified Francoiis Chollet's Jupyter notebook 6.3 from his book "Deep Learning with Python" so that it runs on tensorflow.keras rather than keras (i.e. "from tensorflow.keras import layers" instead of "from keras import layers" etc)
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- TensorFlow installed from (source or binary): built from source
- TensorFlow version (use command below): 2.0.0 (i.e. the release)
- Python version: 3.7 conda
- Bazel version (if compiling from source): 0.26.1
- GCC/Compiler version (if compiling from source): 7.4.0
- CUDA/cuDNN version: 10 / 7.6.4
- GPU model and memory: RTX 2080 Ti and Tesla V100 (tried on both. error occurs on both)

**Describe the current behavior**
Please see attached Jupyter Notebook 
[6.3-advanced-usage-of-recurrent-neural-networks-Copy1.zip](https://github.com/tensorflow/tensorflow/files/3680903/6.3-advanced-usage-of-recurrent-neural-networks-Copy1.zip).  I am going through Francois Chollet's book "Deep Learning with Python" and running the code in his Jupyter Notebooks in Tensorflow 2.0.0 adapting the code to "import tensorflow.keras" instead of "import keras".  Notebook 6.3, (under the heading "1.6 Using recurrent dropout to fight overfitting") has a model with a tensorflow.keras.layers.GRU(32, dropout=0.2, recurrent_dropout=0.2, input_shape=(None, float_data.shape[-1])).  The data is read earlier in the notebook from jena_climate_2009_2016.csv. With tensorflow 2.0.0 and tensorflow.keras I get a loss of 20417499919998144512.0000 and val_loss: 1.1059 after the first epoch and similar figures after subsequent epochs.  These figures are simply wrong (see below).  It also runs about 10x slower than its supposed to.  I interrupted the kernel after 4 epochs.  The original notebook (from Francois Chollet) is here: [link to github](https://github.com/fchollet/deep-learning-with-python-notebooks/blob/master/6.3-advanced-usage-of-recurrent-neural-networks.ipynb) and includes the correct output.

**Describe the expected behavior**
I ran the same code with keras (not tf.keras) using a tensorflow 1 backend a while ago and it gave correct results.  The loss after 1 or 2 epochs is supposed to be around 0.3  The validation loss is supposed to be a little less than 0.3.  The graph below this code was produced using keras and a tensorflow 1 backend.  It shows the correct output.  Francis Chollet's original notebook (as linked to github above) also shows the correct output.

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.
Download the data as follows:
cd ~
mkdir Datasets
cd ~/Datasets
mkdir jena_climate
cd jena_climate
wget https://s3.amazonaws.com/keras-datasets/jena_climate_2009_2016.csv.zip
unzip jena_climate_2009_2016.csv.zip

Run jupyter notebook and load the notebook that is attached to this issue in a Python 3.7 environment with tensorflow 2.0.0.
In the notebook replace /home/daniel with /home/(your username).
Run each cell from the beginning of the notebook so you load the data and create the generators before you get to the example under heading 1.6.  Then try to run the example.  You will find that the loss and val_loss are terribly wrong.

EDIT: Since writing this, I have tried to run more code in the notebook.  The code under heading "1.7 Stacking recurrent layers" also runs incorrectly in tensorflow 2.0.0 using tensorflow.keras.  The loss produced is "nan" (it should be around 0.3).  I think it is the same problem with layers.GRU

The problem does **not** occur with tensorflow.keras in tensorflow 1.1.4.

EDIT: The rest of the remaining code in the notebook runs correctly, e.g. bidirectional GRU runs OK

**Comment by qlzh727** [comment]
Thanks for reporting the issue. I can reproduce it locally. It seems that only when adding recurrent_dropout, the loss just peaked. I will take a closer look wrt to the recurrent_dropout.

**Comment by qlzh727** [comment]
Checked recurrent_dropout in details, it seems that the issue only occurs in implementation=2 which is the default in tf 2.0 (we changed from implementation 1 to 2 as default value). In short, the implementation 1 and 2 should be numerically same, but just be implemented differently. 1 uses more samll tensors and slide weights into individual gates, while 2 uses more bulk matmul. 

The issue I believe is caused by an oversight when implementation 2 is introduce. When dropout is enabled, we applied the same dropout masks for all the gates, since it doesn't slide weights into individual gates. This might cause some numerical instability, eg if both candidate gate and forget gate gets the same masked state, they might be contradictory to each other (I don't have the detailed math proof here).

In short, for now you can build GRU with "implementation=1" in the init arg to walk around issue. I will send a fix very soon to force the implementation == 1 if the recurrent_dropout is specified.

**Comment by qlzh727** [comment]
Should be fixed now.

**Comment by dbonner** [comment]
Hi,
Was this issue fixed on master or r2.0 branch or both?
Cheers,
Daniel

**Comment by qlzh727** [comment]
This is only fixed in HEAD so nightly will have it, but not 2.0. We will have tf 2.1 soon which will contain this fix.

**Comment by ajyeager** [comment]
Hi, is there any more information on what the issue was (conceptual bug vs numerical instability)?  We've been using implementation=2 on the LSTM and recurrent_dropout for a while without issue.  We found this code change when debugging a 400% increase in inference time for identically defined Kera's models and are considering ways to force implementation 2 (downgrade to TF2.0 before this fix or remove recurrent_dropout).
